### ISSUE TITLE
Preprocessing Bugs

### ISSUE THREAD

**Issue by roheth286** [csivitu/CSI_ForkThis25_Clustering#1]
Section 1.2 discusses common preprocessing tips and techniques, but the implementation is incorrect. Identify the errors and correct them.

Prerequisite: None

**Comment by JMANISHA1304** [comment]
1. PCA Components Selection:
Problem: PCA with fixed n_components may not capture enough variance or may include too much noise.
Solution: Use explained variance ratio (np.cumsum(pca.explained_variance_ratio_)) and choose components until ~90–95% variance is retained.
Example:
pca = PCA(n_components=0.95, random_state=42)  # auto-selects number
X_pca = pca.fit_transform(X_scaled)
Visualize scree plot to justify your choice.

2. Cluster Number Determination
Problem: KMeans/GMM requires k, but it’s not clear how to choose.
Solution: Run Elbow Method + Silhouette Analysis together. For imbalanced data, also check Calinski-Harabasz and Davies-Bouldin. If possible, compare clustering results to the true labels (Adjusted Rand Index).

3. Evaluation Metrics
Problem: Only one metric (e.g., silhouette) is misleading.
Solution: Use multiple internal metrics: Silhouette, Calinski-Harabasz, Davies-Bouldin. If labels exist: Adjusted Rand Index (ARI), Normalized Mutual Info (NMI), confusion matrix.

4. Visualization Clarity
Problem: PCA/t-SNE plots may be confusing without clear legends.
Solution: Always add axis labels: “PC1 (25% variance)” etc. Add colorbars/legends (cluster vs ground truth). For high sample sizes: scatter density (hexbin) instead of raw scatter.
